docs/adr/adr-013-automatic-gas-price-estimation.md

```diff
@@ -0,0 +1,65 @@
+# ADR #013: Automatic Gas Price Estimation
+
+## Changelog
+
+- 17.01.2024: Initial design
+
+## Status
+
+Proposed
+
+## Context
+
+[CIP 18](https://github.com/celestiaorg/CIPs/blob/main/cips/cip-18.md) defines an interface that a light node can use for transaction submission to retrieve an estimate of the gas price and amount of gas required to include their transaction on Celestia. This predominantly is to allow for greater dynamism during times of congestions where gas prices become greater than the global minimum.
+
+This ADR defines how the go light node will interact with the service as a client. Note this does not cover how the server will actually estimate the gas prices and gas required.
+
+## Decision
+
+Allow users of Celestia nodes to opt in to automatic gas price and gas usage estimation for every write method
+
+## Detailed Design
+
+### When to estimate the gas price
+
+Users can opt in to automatic price estimation by using the existing `TxConfig.isGasPriceSet`. Previously, if this was false, the `DefaultGasPrice` was used, now if a user isn't manually setting the gas price, the node will query it from the `GasEstimator` service.
+
+### When to estimate the gas required
+
+Depending on the message types submitted, the node will decide whether to also query for the amount of gas that is required to submit the transaction. In the case of the following messages, the gas can be reliably estimated using a simple linear model:
+
+- `MsgPayForBlob`
+
+For all other messages (or combination of messages), the node will call the `GasEstimator` service.
+
+### Setting guardrails
+
+A new field in the `TxConfig` will be introduced allowing for a user to  use the gas price estimation but setting a maximum gas price as a cap that the user is willing to pay. If the price returned by the estimator is greater than the cap, likely due to a price spike, an error will be returned informing the user that the gas price required is currently greater than the user is willing to pay
+
+### Defaults
+
+By default, the same consensus node that the node is connected with, should be used as the `GasEstimator`. During construction, a user can optionally provide a grpc address that can be used instead.
+
+### Fallback
+
+If there are any errors with interacting with the `GasEstimator` service, the node should log an error and fallback to using the minimum gas price. 
```

Comment:
Should you propagate this error back to the user somehow? Otherwise caller will have no idea why their txs might be taking relatively long.

Reply:
Are there negative financial consequences for doing this fallback? Is it possible that by looping in fallback we spend much more gas then expected? If that's possible, I think we should error back to user.

Reply:
It's not an error to take a long time. There are other ways a user can view the progress of their transaction and with something like replace-by-fee in the future, they will even be able to boost the gas price of their transaction

Reply:
But utilising the fallback is unexpected behaviour. I guess maybe if it's well documented enough how the estimator works then no error returned is fine.

Reply:
Yeah I don't think it's unexpected if it's specified but I was conscious that it may be better to fail loudly if the estimation service is unavailable.

An idea could be, in the manual submission path, allow for someone to specify the minimum that way users can easily encode that fallback.

Reply:
What if the error occurs after submitting the transaction( e.g `out of gas`)? Should also fall back to a default solution with an additional warning to the user(in case he provides his own estimator service address)?

Reply:
Let's clarify this please @cmwaters

Reply:
I've come back around on this and think that we should remove the fallback. It should just error to the user in the case that their service is unavailable.